File: ch03/18/HasSubtree.py
# -*- coding:utf-8 -*-
# 输入两棵二叉树A，B，判断B是不是A的子结构。（ps：我们约定空树不是任意一个树的子结构）
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        # write code here
        result = False
        if pRoot1 != None and pRoot2 != None:
            if pRoot1.val == pRoot2.val:
                result = self.DoesTree1HasTree2(pRoot1, pRoot2) 
            if not result:
                result = self.HasSubtree(pRoot1.left, pRoot2)
            if not result:
                result = self.HasSubtree(pRoot1.right, pRoot2)
        return result

    def DoesTree1HasTree2(self, p1, p2):
        if p2 == None:
            return True
        if p1 == None:
            return False
        if p1.val != p2.val:
            return False

        return self.DoesTree1HasTree2(p1.left, p2.left) and \
               self.DoesTree1HasTree2(p1.right, p2.right)

File: ch03/18/HasSubtree.cpp
// 输入两棵二叉树A，B，判断B是不是A的子结构。（ps：我们约定空树不是任意一个树的子结构）
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    //第二遍刷题
    bool doesTree1HaveTree2(TreeNode* pRoot1, TreeNode* pRoot2) {
        if (pRoot2 == NULL)
            return true;
        
        if (pRoot1 == NULL)
            return false;
        
        if (pRoot1 -> val != pRoot2 -> val)
            return false;
        
        return doesTree1HaveTree2(pRoot1 -> left, pRoot2 -> left) && doesTree1HaveTree2(pRoot1 -> right, pRoot2 ->right);
    }
    
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        bool result = false;
        if (pRoot1 == NULL || pRoot2 == NULL)
            return result;
        
        if (pRoot1 != NULL && pRoot2 != NULL) {
            //前序遍历找到tree1等于tree2根节点的节点
            if (pRoot1 -> val == pRoot2 -> val)
                result = doesTree1HaveTree2(pRoot1, pRoot2);
            
            if (!result)
                result = HasSubtree(pRoot1->left, pRoot2);
            
            if (!result)
                result = HasSubtree(pRoot1->right, pRoot2);
        }
        
		return result;
    }
};